server-plugins/ai-bot-resources/src/index.ts

```diff
@@ -80,44 +78,88 @@ async function OnMessageSend (originTxs: TxCreateDoc<ChatMessage>[], control: Tr
     await control.findAll(control.ctx, contact.class.SocialIdentity, { key: aiBotEmailSocialKey }, { limit: 1 })
   )[0]
+}
 
-  if (primaryIdentity === undefined) return []
+async function findAiBotAllSocialIds (control: TriggerControl, socialIdentity: SocialIdentity): Promise<PersonId[]> {
+  const { account } = control.ctx.contextData
+  if (account.socialIds.includes(socialIdentity._id)) {
+    return account.socialIds
+  }
 
-  const allAiSocialIds: PersonId[] = account.socialIds.includes(primaryIdentity._id)
-    ? account.socialIds
-    : (
-        await control.findAll(control.ctx, contact.class.SocialIdentity, {
-          attachedTo: primaryIdentity.attachedTo
-        })
-      ).map((it) => it._id)
+  return (
+    await control.findAll(control.ctx, contact.class.SocialIdentity, { attachedTo: socialIdentity.attachedTo })
+  ).map((it) => it._id)
+}
 
-  const { hierarchy } = control
-  const txes = originTxs.filter((it) => !allAiSocialIds.includes(it.modifiedBy))
+function isThreadMessage (control: TriggerControl, message: ChatMessage): boolean {
+  return control.hierarchy.isDerived(message._class, chunter.class.ThreadMessage)
+}
 
-  if (txes.length === 0) {
-    return []
+async function getMessageDoc (message: ChatMessage, control: TriggerControl): Promise<Doc | undefined> {
+  if (isThreadMessage(control, message)) {
+    const thread = message as ThreadMessage
+    const _id = thread.objectId
+    const _class = thread.objectClass
+
+    return (await control.findAll(control.ctx, _class, { _id }))[0]
   }
+  const _id = message.attachedTo
+  const _class = message.attachedToClass
 
-  for (const tx of txes) {
-    const message = TxProcessor.createDoc2Doc(tx)
+  return (await control.findAll(control.ctx, _class, { _id }))[0]
+}
+
+async function isDirectAvailable (direct: DirectMessage, control: TriggerControl): Promise<boolean> {
+  const account = await getAccountBySocialKey(control, aiBotEmailSocialKey)
+  if (account == null) {
+    return false
+  }
 
-    const isThread = hierarchy.isDerived(tx.objectClass, chunter.class.ThreadMessage)
-    const docClass = isThread ? (message as ThreadMessage).objectClass : message.attachedToClass
+  return direct.members.length === 2 && direct.members.includes(account)
+}
 
-    if (!hierarchy.isDerived(docClass, chunter.class.DirectMessage)) {
-      continue
-    }
+async function isAiBotShouldReply (
+  control: TriggerControl,
+  message: ChatMessage,
+  messageDoc: Doc,
+  aiBotSocialIdentity: SocialIdentity
+): Promise<boolean> {
+  const aiBotPersonId = aiBotSocialIdentity.attachedTo
 
-    if (docClass === chunter.class.DirectMessage) {
-      await onBotDirectMessageSend(control, message)
-    }
-  }
+  const isDirect =
+    messageDoc._class === chunter.class.DirectMessage && (await isDirectAvailable(messageDoc as DirectMessage, control))
+  const isAiBotPersonalChat = messageDoc._id === aiBotPersonId
+  const isAiBotMentioned = message.message.includes(aiBotPersonId)
```

Comment:
It's not enough to check just `includes(id)`. We need to ensure there's a mention `node` and inspect its attributes to confirm it's indeed referencing `ai`. You can look at how `isDocMentioned` works in `server-plugins/activity-resources/src/references.ts.`

Reply:
Thank you, refactored with honest checking in references.